File: algorithms-week8/strongly_connected/python/strongly_connected.py
# Uses python3

import sys

sys.setrecursionlimit(200000)


def topological(digraph):
    n = len(digraph)
    visited = [False] * n
    order = []

    for v in range(n):
        if visited[v] == False:
            dfs(digraph, visited, v, order)

    return reversed(order)


def dfs(adj, visited, v, order=None):
    visited[v] = True

    for w in adj[v]:
        if visited[w] == False:
            dfs(adj, visited, w, order)

    if order is not None:
        order.append(v)


def number_of_strongly_connected_components(adj):
    result = 0
    n = len(adj)
    visited = [False] * n
    reverse = [[] for _ in range(n)]
    for v in range(n):
        for w in adj[v]:
            reverse[w].append(v)

    topo = topological(reverse)

    for v in topo:
        if visited[v] == False:
            dfs(adj, visited, v)
            result += 1

    return result


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    print(number_of_strongly_connected_components(adj))

File: algorithms-week8/strongly_connected/cpp/strongly_connected.cpp
//
// Created by Sam on 9/11/2018.
//

#include <algorithm>
#include <iostream>
#include <vector>

using std::vector;
using std::pair;

static void dfs(vector<vector<int>>& adj, vector<bool>& visited, int v, vector<int>& order) {
  visited[v] = true;

  for (int& w : adj[v]) {
    if (!visited[w]) {
      dfs(adj, visited, w, order);
    }
  }

  order.push_back(v);
}

static void dfs(vector<vector<int>>& adj, vector<bool>& visited, int v) {
  visited[v] = true;

  for (int& w : adj[v]) {
    if (!visited[w]) {
      dfs(adj, visited, w);
    }
  }
}

static vector<int> topological(vector<vector<int>>& digraph) {
  size_t n = digraph.size();
  vector<int> order(n);
  vector<bool> visited(n);

  for (int v = 0; v < n; v++) {
    if (!visited[v]) {
      dfs(digraph, visited, v, order);
    }
  }

  std::reverse(order.begin(), order.end());
  return order;
}

int number_of_strongly_connected_components(vector<vector<int> > adj) {
  int result = 0;
  size_t n = adj.size();
  vector<bool> visited(n, false);
  vector<vector<int>> reverseAdj(n, vector<int>());
  for (int v = 0; v < n; v++) {
    for (int w : adj[v]) {
      reverseAdj[w].push_back(v);
    }
  }

  vector<int> topo = topological(reverseAdj);
  for (int& v : topo) {
    if (!visited[v]) {
      dfs(adj, visited, v);
      result++;
    }
  }
  return result;
}

int main() {
  size_t n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
  }
  std::cout << number_of_strongly_connected_components(adj);
}
